src/PFire.Core/Protocol/Messages/Outbound/ChatroomJoinInfo.cs

```diff
@@ -0,0 +1,143 @@
+
+            if (Response == (int)ChatResponse.Fail) //Are we already invalid? Then silently fail.
+            {
+                IsSilenced = 0;
+                IsPassworded = 0;
+                ResponsePermission = 0;
+            }
+            else if (_isNewRoom)
+            {
+                ResponsePermission = (int)Permissions.Admin;
+                IsSilenced = 0;
+                IsPassworded = 0;
+            }
+            else
+            {
+                var userLevel = await context.Server.Database.GetUserHighestChatPowerLevel(_chatroomPayload.Id, context.User.Id);
+
+                //Check to see if Default Permissions are set and see if the user has a non-standard permission already.
+                if (_chatroomPayload.DefaultPerms != 2 && userLevel == (int)Permissions.Normal)
+                {
+                    await context.Server.Database.UpdateChatroomUserPerms(_chatroomPayload.Id, context.User.Id, _chatroomPayload.DefaultPerms);
+                    ResponsePermission = _chatroomPayload.DefaultPerms;
+                }
+                else
+                {
+                    ResponsePermission = userLevel;
+                }
+
+                IsSilenced = _chatroomPayload.Silenced;
+
+                IsPassworded = (byte)(!string.IsNullOrEmpty(_chatroomPayload.Password) ? 1 : 0);
+                if (IsPassworded == 1 && string.IsNullOrEmpty(_password))
+                {
+                    Response = (int)ChatResponse.PasswordRequired; //Password is needed
+                } 
+                else if (IsPassworded == 1 && !string.IsNullOrEmpty(_password))
+                {
+                    Response = (byte)(_password == _chatroomPayload.Password ? (int)ChatResponse.OK : (int)ChatResponse.WrongPassword); //Wrong Password, try again.
+                }
+            }
+
+            //We all good? Send it to everyone that we have entered.
+            if (Response == (int)ChatResponse.OK)
+            {
+                await context.Server.Database.UserEnteredRoom(ChatId, context.User.Id);
+                
+                if (_chatroomPayload.GameLobbyID != 0)
+                {
+                    await context.SendAndProcessMessage(new ChatroomGameLobbyInfo(ChatId, _chatroomPayload.GameLobbyID, _chatroomPayload.GameLobbyHost, _chatroomPayload.GameLobbyIP, _chatroomPayload.GameLobbyPort, _chatroomPayload.GameLobbyPlayers));
+                }
+
+                var perms = await context.Server.Database.GetUserHighestChatPowerLevel(ChatId, context.User.Id);
+                var users = await context.Server.Database.QueryUsersInChat(ChatId);
+                foreach (var user in users)
+                {
+                    var friendClient = context.Server.GetSession(user);
+                    if (friendClient != null)
+                    {
+                        await friendClient.SendAndProcessMessage(new ChatroomUserJoined(ChatId, context.User, perms));
```

Comment:
It's confusing to me that this outbound message is creating more messages. What triggers this message? We should try to send these messages from there instead.

Ideally the outbound messages are doing one thing and processing _THIS_ message.

Reply:
Join Info is the entire process of attempting to join a chatroom. From checking to see if it's locked down, to password entry to denying for various factors. 

The message that gets sent, User joined is telling the other people in the chatroom that you joined if you passed all the checks and you join the chatroom.

If you want to, say keep the ethos of one packet per process, we could maybe return a successful join or something.

Reply:
Hmm let me think about this one more. Maybe I'll get a better understanding of what we can do with it as I read more of the code